### ISSUE TITLE
Best way to handle state

### ISSUE THREAD

**Issue by javivelasco** [react-toolbox/react-toolbox#800]
Hi all! sorry for the accumulation of PRs, I'm merging and checking stuff during the weekend 🙏  #774 made me think about how we handle state and I want to throw a question to decide in which direction should we move. 

Right now we are keeping the state for components like `Input` **outside** components because in most use cases you'd keep control of the state outside the component so using the component standalone is just not possible. The same happens for other components similar to inputs but this is not the case for `Dropdown` for example.

In `Dropdown` we are assuming most of the time you'd want to render it closed, close it when a click outside happens and open it with a click inside the internal Input. But what if you want to keep the state out or close the component from outside? Maybe stripping out the state is more flexible but also more painful in most use cases. A solution for this is to implement both: we can still keep an internal state but also synchronize an external `active` prop on `componentWillReceiveProps` lifecycle and notify parents about an internal state change with a property `onActiveChange`. There is of course a third state type that only makes sense internally but notice I'm referring to state likely to be used from outside only.

I'm aware it's really easy to add state to a stateless component, for example, decorating using `withState` from [recompose](https://github.com/acdlite/recompose). My personal opinion is that the best would be to keep every component written in an stateless fancy so if you want to add state you just decorate it. But this should be done *outside* the library cause it's messy enough with styled/unstyled imports. So, two options:

- Every component should have state control outside. Consumers that want to use a component as standalone should wrap or decorate it.
- Every component should have state control inside but also provide props to synchronize and delegate control from outside.

What do you think we should do? The most immediate would be extracting state in the Dropdown vs adding props to sync state.

**Comment by acornejo** [comment]
I think that going stateless is best for most components, but there are a few exceptions where convenience trumps purity.

In this case we might get the best of both worlds by having two components. 1) A stateless dropdown component much like the one today with a couple of additional props to control if its active or not. 2) A select component which is internally implemented using  the dropdown component, and which offers a much more restricted API but requires no state management.

If you are using the full power of dropdown and leveraging the capability of custom renderer methods, etc, then it is not a lot of added burden to ask the user to manage the active state as well. On the other hand, if you just want a the user to be able to select between "medium,small, large"  with no fancy icons then you can use a select field and be done with it.

The alternative you propose of providing a single component that provides both API's at the same time (you manage state outside, or it manages state inside) seems like it would be slightly harder to maintain and debug.

**Comment by therealparmesh** [comment]
I definitely believe in the stateless approach. In the project I'm working on, we let redux control almost everything because of complex dependencies in our business logic. So for us, it's a necessity that we avoid statefulness as much as we can.

I ended up writing my own more-or-less stateless Autocomplete equivalents, a Select and a MultiSelect, using other components and SASS themes in this library. (I say more-or-less because there is a little bit of local state for logic related to the display.) We decided to do this because we love the material look-and-feel of this library, but needed the `source` (we called it `store`) and `value` properties to be implemented in a stateless manner. We also ended up adding a `source`-like property (we called it `displayStore`) that we use purely to determine what value is shown in the Input and that we don't show in the dropdown list. I imagine that if we require the Dropdown component anytime soon, we would take the same approach.

If this becomes a priority for react-toolbox, I'd love to take part in the solution given my team's real-world-based needs for stateless UI components.

**Comment by javivelasco** [comment]
Great, for now I'm in on rewriting components in a totally stateless fancy adding local state with recompose where needed. Let's keep this open to gather opinions, still there are some priorities to complete before we can really afford this.

**Comment by omarstreak** [comment]
I think having 2 components, 1 stateless and 1 with state is the best way to go.

Lots of complex apps usually need some kind of custom behavior so the less you're fighting with the UI framework the better. The more the component does the higher likelihood you'll be fighting with the component. So providing a simple stateless component that does one thing and does it well means that it can compose easily. Then providing a more advanced component for basic use cases built on top gives you the best of both worlds.

**Comment by borela** [comment]
The components should handle the state themselves, this is a lesson I learned back from the Winforms/WPF era, your view model must be pure, it shouldn't care if your dropdown is open or not or if it is a dropdown at all... all it needs is which item from a dataset the user selected and if the user is able to select it or not.

For me, the #774 raises more questions:

- Why would need to close it when something happen in another component?
- Does the dataset become stale? Then just update it.
- Is the user not allowed to select the item anymore? Then disable it.

**Comment by javivelasco** [comment]
@borela Sorry about this delayed answer. I've been out for vacation, I really needed :) 

In my opinion the best is to manage the UI in an stateless way. Then we can use something like `recompose` to add state to those components. In my experience it makes easier the development and testing. This means that optionally you could import a component without any state gaining flexibility, but by default the component is encapsulated with state.... but it's handled outside the component. 

Does this makes sense?

About #774. It's a very common behavior to close a Dropdown on a click outside regardless it concerns internal state. I think that having an internal state that gets modified on an click outside the component just means that there should be a communication between the component and the outside world somehow. With an implementation that extract the state it should be very easy to implement.

**Comment by borela** [comment]
@javivelasco I understand that the flexibility seams appealing for testing, but to me the UI shouldn't care if the combobox is open or not that's a behaviour exclusive for the control, all it needs to care as if the dataset has changed and if it did, then the combobox would update itself automatically.

**Comment by javivelasco** [comment]
Yeah, agree in most scenarios. The approach of writing component stateless and providing the state with HOC would still make it work like it is now but it's more flexible because it simplifies testing and allows you to import the component and take care of the state your self. The default would be still with state though
